apps/managedidentity/managedidentity.go

```diff
@@ -204,28 +292,183 @@ func (client Client) getTokenForRequest(req *http.Request) (accesstokens.TokenRe
+	msiParameters.Set(apiVersionQueryParameterName, imdsAPIVersion)
+	resource = strings.TrimSuffix(resource, "/.default")
+	msiParameters.Set(resourceQueryParameterName, resource)
+
+	switch t := id.(type) {
+	case UserAssignedClientID:
+		msiParameters.Set(miQueryParameterClientId, string(t))
+	case UserAssignedResourceID:
+		msiParameters.Set(miQueryParameterResourceId, string(t))
+	case UserAssignedObjectID:
+		msiParameters.Set(miQueryParameterObjectId, string(t))
+	case systemAssignedValue: // not adding anything
+	default:
+		return nil, fmt.Errorf("unsupported type %T", id)
 	}
-	req, err := createIMDSAuthRequest(ctx, client.miType, resource, o.claims)
+
+	msiEndpoint.RawQuery = msiParameters.Encode()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, msiEndpoint.String(), nil)
 	if err != nil {
-		return base.AuthResult{}, err
+		return nil, fmt.Errorf("error creating http request %s", err)
+	}
+	req.Header.Set(metaHTTPHeaderName, "true")
+	return req, nil
+}
+
+func createAzureArcAuthRequest(ctx context.Context, id ID, resource string) (*http.Request, error) {
+	identityEndpoint, _, _ := getAndValidateAzureArcEnvVars()
+	var msiEndpoint *url.URL
+
+	if identityEndpoint == "" {
+		identityEndpoint = azureArcEndpoint
+	}
+
+	if _, ok := id.(systemAssignedValue); !ok {
+		return nil, errors.New("Azure Arc doesn't support user assigned managed identities")
+	}
+
+	identityEndpoint = strings.Replace(identityEndpoint, "localhost", "127.0.0.1", -1)
+	msiEndpoint, parseErr := url.Parse(identityEndpoint)
+	if parseErr != nil {
+		return nil, fmt.Errorf("couldn't parse %q: %s", identityEndpoint, parseErr)
 	}
-	tokenResponse, err := client.getTokenForRequest(req)
+
+	msiParameters := msiEndpoint.Query()
+	msiParameters.Set(apiVersionQueryParameterName, azureArcAPIVersion)
+	resource = strings.TrimSuffix(resource, "/.default")
+	msiParameters.Set(resourceQueryParameterName, resource)
+
+	msiEndpoint.RawQuery = msiParameters.Encode()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, msiEndpoint.String(), nil)
 	if err != nil {
-		return base.AuthResult{}, err
+		return nil, fmt.Errorf("error creating http request %s", err)
 	}
-	return base.NewAuthResult(tokenResponse, shared.Account{})
+	req.Header.Set(metaHTTPHeaderName, "true")
+	return req, nil
+}
+
+// GetEnvironmentVariables returns the identity and IMDS endpoints
+func getAndValidateAzureArcEnvVars() (string, string, bool) {
+	identityEndpoint := os.Getenv(IdentityEndpointEnvVar)
+	imdsEndpoint := os.Getenv(ArcIMDSEnvVar)
+
+	if identityEndpoint != "" && imdsEndpoint != "" {
+		return identityEndpoint, imdsEndpoint, true
+	}
+
+	if identityEndpoint == "" || imdsEndpoint == "" {
+		platform := runtime.GOOS
+
+		// Only platforms inside the azureArcOsToFileMap are supported
+		fileDetectionPath, platformSupported := azureArcOsToFileMap[platform]
+
+		if platformSupported && fileExists(fileDetectionPath) {
+			identityEndpoint = azureArcEndpoint
```

Comment:
returning identity endpoint from this method is all that you need. You can skip returning the imds endpoint. Since at this point you have confirmed that the env is Azure arc and imds endpoint is only used for that purpose.